iree-prof-tools/graph-util.cc

```diff
@@ -0,0 +1,428 @@
+absl::StatusOr<mlir::func::FuncOp> GetFuncWithName(mlir::ModuleOp module,
+                                                   absl::string_view name) {
+  absl::StatusOr<mlir::func::FuncOp> result = absl::NotFoundError(
+      absl::StrCat("Can't find a func of name \"", name, "\""));
+
+  module->walk([&](mlir::func::FuncOp func) -> mlir::WalkResult {
+    if (name == func.getSymName().str()) {
+      result = func;
+      return mlir::WalkResult::skip();
+    }
+    return mlir::WalkResult::advance();
+  });
+
+  if (result.ok()) {
+    LOG(INFO) << "Found a func with name = " << result->getSymName().str();
+  }
+  return result;
+}
+
+// Adds a node with |label| and |parent_namespace| into |graph|.
+GraphNode& AddNode(absl::string_view label,
+                   absl::string_view parent_namespace,
+                   Graph& graph) {
+  graph.nodes.push_back(std::make_unique<GraphNode>());
+  auto& node = *graph.nodes.back();
+  node.id = absl::StrCat(graph.nodes.size() - 1);
+  node.label = label;
+  node.node_namespace = parent_namespace;
+  return node;
+}
+
+// Adds an output of |source| as an incoming edge of |node|.
+IncomingEdge& AddIncomingEdge(const GraphNode& source,
+                              int source_output_index,
+                              GraphNode& node) {
+  IncomingEdge& input = node.incoming_edges.emplace_back();
+  input.source_node_id = source.id;
+  input.source_node_output_id = source.outputs_metadata[source_output_index].id;
+  int input_index = node.incoming_edges.size() - 1;
+  input.target_node_input_id = absl::StrCat("input-", input_index);
+  input.metadata.emplace_back("index", absl::StrCat(input_index));
+  return input;
+}
+
+// Adds an output with some default metadata into |node|.
+Metadata& AddOutputMetadata(GraphNode& node) {
+  Metadata& output = node.outputs_metadata.emplace_back();
+  int output_index = node.outputs_metadata.size() - 1;
+  output.id = absl::StrCat("output-", output_index);
+  output.attrs.emplace_back("index", absl::StrCat(output_index));
+  return output;
+}
+
+// Adds a node for |op| into |graph|.
+GraphNode& AddNodeForOperation(
+    mlir::Operation* op,
+    absl::string_view label,
+    const absl::flat_hash_map<mlir::Operation*, std::string>& op_namespaces,
+    Graph& graph) {
+  CHECK(op_namespaces.contains(op->getParentOp()));
+  return AddNode(label, op_namespaces.at(op->getParentOp()), graph);
+}
+
+// Adds a node for |op| with a default label, i.e. op's name into |graph|.
+GraphNode& AddNodeForOperation(
+    mlir::Operation* op,
+    const absl::flat_hash_map<mlir::Operation*, std::string>& op_namespaces,
+    Graph& graph) {
+  return AddNodeForOperation(op, ToStr(op->getName()), op_namespaces, graph);
+}
+
+// Finds a node whose label is matched with |label_to_match|.
+GraphNode* FindNodeForLabel(const Graph& graph,
+                            absl::string_view label_to_match) {
+  for (const auto& n : graph.nodes) {
```

Comment:
nit: If this can become expensive, we may introduce a class for all APIs here and keep a hash map. (Similarly for `op_namespaces` to simplify the signature of AddNode*().) (No need to change in this PR.)